packages/vite/src/node/assetSource.ts

```diff
@@ -0,0 +1,231 @@
+import type { DefaultTreeAdapterMap, Token } from 'parse5'
+
+interface HtmlAssetSource {
+  tag: string
+  attributes: string[]
+  type: 'src' | 'srcset'
+  filter?: (data: HtmlAssetSourceFilterData) => boolean
+}
+
+interface HtmlAssetSourceFilterData {
+  attribute: string
+  value: string
+  attributes: Record<string, string>
+}
+
+interface AssetAttribute {
+  attribute: Token.Attribute
+  type: 'src' | 'srcset'
+  location: Token.Location
+}
+
+export function getNodeAssetAttributes(
+  node: DefaultTreeAdapterMap['element']
+): AssetAttribute[] {
+  const assetAttrs: AssetAttribute[] = []
+
+  for (const assetSource of DEFAULT_HTML_ASSET_SOURCES) {
+    if (assetSource.tag !== node.nodeName) continue
```

Comment:
Maybe it's more efficient if we changed `DEFAULT_HTML_ASSET_SOURCES` to an object instead of an array?

```js
const assetSource = DEFAULT_HTML_ASSET_SOURCES[node.nodeName]
if (!assetSource) return
```

Reply:
It would be more performant, but if we want to make this extensible by users in the future, it makes the API a bit weird. For example, they have to mutate the `img` key and value to support `data-src`. 🤔

Reply:
Actually we could probably keep an object form internally for perf, and allow an array syntax when we expose an API for it.